IDE/iotsafe-raspberrypi/main.c

```diff
@@ -0,0 +1,288 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+/* IoT-SAFE example
+ * main for POSIX serial communication.
+ */
+
+/* C Standard Library */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+
+
+
+/* POSIX Library */
+
+#include <errno.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <termios.h>
+#include <signal.h>
+
+
+
+/* wolfSSL Library */
+
+#include <wolfssl/options.h>
+#include <wolfssl/wolfcrypt/settings.h>
+#include <wolfssl/ssl.h>
+#include <wolfssl/wolfcrypt/random.h>
+#include <wolfssl/wolfcrypt/port/iotsafe/iotsafe.h>
+
+
+
+/* Variable Declarations */
+
+static int serial_fd = -1;
+
+
+
+/* Function Declarations */
+
+extern int client_loop(const char *peer_ip, const char *peer_name, const char *peer_port, const char *temperature);
+
+static void print_buffer_hex(const char *buf, int len);
+static void print_buffer_char(const char *buf, int len);
+
+static int usart_init(const char *dev_name, int *fd);
+static int usart_clean(int fd);
+static int usart_read(char *buf, int len);
+static int usart_write(const char *buf, int len);
+
+static void show_usage(const char *program);
+
+
+
+/* Function Definitions */
+
+static void print_buffer_hex(const char *buf, int len)
+{
+    for (int i = 0; i < len; i++)
+        printf("%02X", (unsigned int)*(buf++));
+}
+
+static void print_buffer_char(const char *buf, int len)
+{
+    for (int i = 0; i < len; i++)
+        printf("%c", *(buf++));
+}
+
+static int usart_init(const char *dev_name, int *fd)
+{
+    int portfd = open(dev_name, O_RDWR | O_NOCTTY);
+    if (portfd < 0)
+    {
+        *fd = -1;
+        return errno;
+    }
+
+    struct termios tty;
+    tcgetattr(portfd, &tty);
+    cfsetospeed(&tty, B115200);
+    cfsetispeed(&tty, B115200);
+    tty.c_cflag = (tty.c_cflag & ~CSIZE) | (CS8);
+    tty.c_iflag &= ~(IGNBRK | IXON | IXOFF | IXANY| INLCR | ICRNL);
+    tty.c_oflag &= ~OPOST;
+    tty.c_oflag &= ~(ONLCR|OCRNL);
+    tty.c_cflag &= ~(PARENB | PARODD | CSTOPB);
+    tty.c_lflag &= ~(ICANON | ECHO | ECHOE | ISIG);
+    tty.c_iflag &= ~ISTRIP;
+    tty.c_cc[VMIN] = 0;
+    tty.c_cc[VTIME] = 5;
+    tcsetattr(portfd, TCSANOW, &tty);
+
+    *fd = portfd;
+
+    return 0;
+}
+
+static int usart_clean(int fd)
+{
+    return close(fd);
+}
+
+static int usart_read(char *buf, int len)
+{
+    if (!buf || len < 0)
+        return -1;
+
+#ifdef DEBUG_UART_IO
+    printf("UART Read Expected : %d bytes\n", len);
+#endif
+
+    int ret = 0;
+    int i = 0;
+    char c;
+    memset(buf, 0, len);
+
+    // Read 1 byte at one time until *buf is full or a POSIX read error like timeout occurs.
+    do
+    {
+        ret = read(serial_fd, &c, 1U);
```

Comment:
The read must be blocking for now since the key generation takes time as well as computing a signature. If I add a timeout while the applet is computing a key pair it will cause some unexpected results.